### PR TITLE
[mini] lower typechecks later

### PR DESCRIPTION
**Pull request by lewurm**
lower typecheck instructions later in the compilation, so we've a chance to do high-level optimization after `method-to-ir`

### CODE REVIEW

**Comment by lewurm** [comment]
builddeb

**Comment by lewurm** [comment]
builddeb

**Comment by lewurm** [comment]
in terms of performance it doesn't change much (fortunately):
http://xamarin.github.io/benchmarker/front-end/pullrequest.html#id=243
http://xamarin.github.io/benchmarker/front-end/pullrequest.html#id=252

also in terms of compile-time we're good:
http://xamarin.github.io/benchmarker/front-end/pullrequest.html#id=247

**Comment by lewurm** [comment]
build

**Comment by lewurm** [comment]
@vargaz can I have a review for this please?

**Comment by kumpera** [comment]
@lewurm insinst is exactly the sort of thing I was hoping to see improve with this.

Delay expanding all things that introduce more BBs until later.

File: mono/mini/method-to-ir.c [hunk 1]
```diff
@@ -15053,6 +14989,72 @@ mono_spill_global_vars (MonoCompile *cfg, gboolean *need_local_opts)
 	g_free (live_range_end_bb);
 }
 
+static void mono_decompose_typecheck (MonoCompile *cfg, MonoBasicBlock *bb, MonoInst *ins)
```
Comment (kumpera) [inline mono/mini/method-to-ir.c]: Move the whole pass to a separate file.

File: mono/mini/mini.c [hunk 2]
```diff
@@ -3499,6 +3510,11 @@ mini_method_compile (MonoMethod *method, guint32 opts, MonoDomain *domain, JitFl
 	MONO_TIME_TRACK (mono_jit_stats.jit_method_to_ir, i = mono_method_to_ir (cfg, method_to_compile, NULL, NULL, NULL, NULL, 0, FALSE));
 	mono_cfg_dump_ir (cfg, "method-to-ir");
 
+	if (cfg->gdump_ctx != NULL) {
+		mono_insert_nop_in_empty_bb (cfg);
```
Comment (kumpera) [inline mono/mini/mini.c]: Why is this needed?

**Comment by kumpera** [comment]
This patch doesn't handle the case in handle_castclass that calls mini_emit_castclass_inst.

**Comment by lewurm** [comment]
@kumpera: do you mean `handle_ccastclass`/`handle_cisinst`?

File: mono/mini/mini.c (hunk 2 above)
Comment (lewurm) [inline mono/mini/mini.c]: it's a hack.  IGV (the graph visualization tool) doesn't like empty basic blocks.

File: mono/mini/mini.c [hunk 3]
```diff
@@ -3577,6 +3593,11 @@ mini_method_compile (MonoMethod *method, guint32 opts, MonoDomain *domain, JitFl
 		mono_cfg_dump_ir (cfg, "local_cprop");
 	}
 
+	MONO_TIME_TRACK (mono_jit_stats.jit_decompose_typechecks, mono_decompose_typechecks (cfg));
```
Comment (kumpera) [inline mono/mini/mini.c]: We're inlining after the front end and that will change the CFG. Good thing you picked a place before the CFG passes and analysis. We do this before SSA formation, which is ok too.

But, said that, what's the reasoning to have it here? We're losing a lot of optimizations that could take advantage of longer BBs.

We could do it all the way down after DCE and AA. Could this pass introduce issues?

Given DCE is crazy cheap, we can do another DCE after decompose.

File: mono/mini/method-to-ir.c [hunk 4]
```diff
@@ -15053,6 +14989,72 @@ mono_spill_global_vars (MonoCompile *cfg, gboolean *need_local_opts)
...
+			ret = handle_castclass (cfg, klass, source, context_used);
+	}
+	EMIT_NEW_UNALU (cfg, move, OP_MOVE, ins->dreg, ret->dreg);
+
+	g_assert (cfg->cbb->code || first_bb->code);
+	MonoInst *prev = ins->prev;
+	mono_replace_ins (cfg, bb, ins, &prev, first_bb, cfg->cbb);
+}
+
+void mono_decompose_typechecks (MonoCompile *cfg)
```
Comment (kumpera) [inline mono/mini/method-to-ir.c]: This pass should be conditional like simd and array ops.

When emitting the IR, set a cfg flag saying that the IR has at least one OP_ISINST or OP_CASTCLASS.

Then condition the whole pass on that flag being set.

**Comment by kumpera** [comment]
@lewurm yeah, all the very branchy code in handle_ccastclass, handle_cisinst, mini_emit_castclass_inst and friends.

File: mono/mini/method-to-ir.c (hunk 1 above)
Comment (lewurm) [inline mono/mini/method-to-ir.c]: I'll do that in another PR. Most util functions used here are `static`, it will require a bit more moving code around.

File: mono/mini/mini.c (hunk 3 above)
Comment (lewurm) [inline mono/mini/mini.c]: Yes, we should do it as late as possible.  IIRC there're some issues around inserting new basic blocks after `mono_optimize_branches`, I'll look into that again.

File: mono/mini/method-to-ir.c (hunk 4 above)
Comment (lewurm) [inline mono/mini/method-to-ir.c]: good point, will do!

File: mono/mini/method-to-ir.c (hunk 1 above)
Comment (vargaz) [inline mono/mini/method-to-ir.c]: There is already a decompose.c file where most of this stuff lives.

File: mono/mini/mini.c (hunk 2 above)
Comment (evincarofautumn) [inline mono/mini/mini.c]: You should probably add a comment about that.

**Comment by lewurm** [comment]
@kumpera so, here's the thing with `CEE_MONO_CCASTCLASS`/`CEE_MONO_CISINST`: both are introduced by decomposing `OP_CASTCLASS` and `OP_ISINST` respectively in order to perform a proxy check. Since we lower those operations in a later stage now, we inline the wrappers containing the `C{CASTCLASS,ISINST}` instructions, however our inliner is basically just `method-to-ir` and therefore those instructions wouldn't be lowered if we'd introduce something like `OP_CCASTCLASS`/`OP_CISINST`.

Possible solutions:

1. Have another phase after `decompose_typechecks` that handles those proxy instructions.
2. Make `decompose_typechecks` iterative.
3. Introduce stages for the compilation pipeline, and make the inliner so that it can lower an inlinee down to a required stage.
4. Leave it as is, and do the decomposition for `C{CASTCLASS,ISINST}` in `method-to-ir`

As for now, I think we should stay with `(4)`. For the long-term, we should look into `(3)`, as this is also required for late inlining.

File: mono/mini/method-to-ir.c (hunk 4 above)
Comment (lewurm) [inline mono/mini/method-to-ir.c]: 3ddbaf012974558fd64cc9403bfedadc0070b7d2

File: mono/mini/mini.c (hunk 2 above)
Comment (lewurm) [inline mono/mini/mini.c]: 41b187035e83dc4f6846f47e0186fa2ad4a3705c

**Comment by kumpera** [comment]
This is so confusing.

Let me see if I understand correctly.

Those special OPs, that expand into huge amounts of IR, only happen in the we wrappers.

The type test wrappers are always inlined, so they are not really needed anywhere else.

Does this 2 stage wrapper + custom OPs makes sense?

It's definitely the case that it's impossible to decipher that behavior for the mess that gsharing introduced.

**Comment by lewurm** [comment]
let me try to dechiper this thing a bit more. For the current PR, it looks like this:

thing | description | 
--- | --- |
`CEE_ISINST` | CLR opcode |
`OP_ISINST` | opcode introduced by `method-to-ir` for `CEE_ISINST`
`decompose_typecheck` | replaces (among other things) `OP_ISINST` with a managed wrapper (generated by `mono_marshal_get_isinst` in `mono/metaclass/marshal.c`). We push the managed wrapper through `method-to-ir`, and replace the resulting CFG with `OP_ISINST` (aka. inlining). 
`CEE_MONO_CISINST` | this opcode is introduced by the above mentioned managed wrapper. `method-to-ir` lowers it directly to mini IR (see `handle_cisinst`) as opposed to another managed code indirection.

I'm not sure why `CEE_ISINST` is lowered via a managed code snippet, but `CEE_MONO_CISINST` isn't. There might be good reasons for that, but I haven't figured them out.

**Comment by lewurm** [comment]
@monojenkins merge
